bin/list-all

```diff
@@ -1,19 +1,97 @@
 #!/usr/bin/env bash
 
-releases_path="https://repo.hex.pm/builds/elixir/builds.txt"
-cmd="curl -s $releases_path"
-
-# stolen from https://github.com/rbenv/ruby-build/pull/631/files#diff-fdcfb8a18714b33b07529b7d02b54f1dR942
-# and adapted
-function sort_versions() {
-  sed 'h; s/.p\([[:digit:]]\)/.z\1/; G; s/\n/ /' | \
-    LC_ALL=C sort -t. -k 1,1 -k 2,2n -k 3,3n -k 4,4n -k 5,5n | awk '{print $2}'
+
+# Download the list of versions available on Hex.pm.
+#
+# Hex.pm keeps a public builds.txt file where it lists all the builds available
+# in their servers.
+#
+# If we're able to successfuly fetch the builds file, its content is written to
+# STDOUT. Otherwise, nothing is written to STDOUT and an error message is
+# written to STDERR.
+download_hex_pm_builds() {
+  local releases_url='https://repo.hex.pm/builds/elixir/builds.txt'
+
+  if ! curl -fs $releases_url; then
+    cat >&2 <<EOS
+==> Failed to fetch Hex.pm versions list.
+
+Hex.pm returned an error for the following URL:
+
+${releases_url}
+
+Make sure your internet connection is stable and that you can reach
+https://hex.pm/.
+
+If the problem persists, please report it on GitHub:
+https://github.com/asdf-vm/asdf-elixir/issues/new
+EOS
+
+    return 1
+  fi
+}
+
+
+get_hex_pm_versions() {
+  # Get the builds file from Hex.pm
+  download_hex_pm_builds |
+    # Filter only the first field (the version name)
+    cut -d\  -f1 |
+    # Normalize version names
+    sed -E '
+      # Discard numeric versions just with major
+      /^v[0-9]+(-.*)?$/d;
+
+      # Discard numeric versions just with major and minor
+      /^v[0-9]+.[0-9]+(-.*)?$/d;
+
+      # Remove `v` prefix from numeric versions
+      s/^v([0-9]+.[0-9]+.[0-9]+(-.*)?)$/\1/
+    '
```

Comment:
Much nicer than the old code! And one sed process instead of two with awk and sed!

Reply:
Your comment actually showed a mistake here, I forgot to properly escape the dots when I changed to extended regex :grimacing:. Changed in 6e60f32.